### ISSUE TITLE
Inquiry regarding Input Processing Flexibility and General Multimodal Reasoning Capabilities

### ISSUE THREAD

**Issue by Yanyx44** [jzhzhang/Uni-NaVid#19]
Thank you for your excellent work on UniNaVid!

I have been studying the `UniNaVid_Agent` code and currently have some questions regarding the working mechanism and input processing logic of the model.

### Observations

Based on my understanding of the current codebase, it seems that input images are strictly treated as **historical trajectory frames**. The processing logic appears to be somewhat coupled with this assumption:

1.  **Frame Processing:** In the `process_images` function, the line `self.model.get_model().new_frames = len(rgb_list)` suggests that the model's internal state update relies heavily on the count of incoming frames as history which affects how input frames are processed.
2.  **Prompt Construction:** In `predict_inference`, specific tokens like `[Navigation]` are automatically appended to the prompt to control action generation.

### The Question

I am concerned that this rigid input structure might limit the model's flexibility for other types of inputs or tasks. Specifically:

* **Flexible Navigation (e.g., Image-Goal Nav):** I aim to perform tasks where the goal is defined by an image (Instance ImageNav). In the current pipeline, simply appending a target image to the input list causes it to be treated as part of the historical trajectory (due to the `new_frames` logic). 
**How can I properly introduce an extra goal image as a distinct input, ensuring it is recognized as the target condition rather than a past observation frame?**
* **General Reasoning:** If I need to feed the model some images for simple visual reasoning (VQA) rather than generating navigation actions, is the current inference code suitable?

It seems the current `[Navigation]` token and the trajectory-based processing might bias the model towards generating actions based on history, rather than analyzing a static image or a goal image.

### Request

Could you provide some guidance or clarification on:
1.  How can we support more flexible multimodal inputs and outputs, rather than the current setup that treats inputs strictly as historical frames and outputs solely as navigation
2.  How to utilize the UniNaVid model for more **general multimodal reasoning** tasks without triggering the specific navigation history logic?

Any advice or code snippets on how to modify the prompt/input formatting for these use cases would be greatly appreciated.

Thanks!

**Comment by jzhzhang** [comment]
Thanks for your interest in our project!

First, all information, including images and text, is organized as tokens. By structuring the tokens (visual tokens and language tokens), you can accomplish other navigation tasks, such as image-goal navigation.

However, these modifications may lead to noticeable differences in token patterns, meaning you would need to retrain Uni-NaVid with your new data. This would also require substantial code modifications (e.g., to the data loader, navid_arch, and other components).

The most related part of the code can be found in:
https://github.com/jzhzhang/Uni-NaVid/blob/64444eba4d4503d5b08f080673e87229f7fe7f25/uninavid/model/uninavid_arch.py#L502-L519

You can reorganize the language and visual tokens here, but please note that this requires other related code modifications accordingly.

**Comment by Yanyx44** [comment]
> Thanks for your interest in our project!
> 
> First, all information, including images and text, is organized as tokens. By structuring the tokens (visual tokens and language tokens), you can accomplish other navigation tasks, such as image-goal navigation.
> 
> However, these modifications may lead to noticeable differences in token patterns, meaning you would need to retrain Uni-NaVid with your new data. This would also require substantial code modifications (e.g., to the data loader, navid_arch, and other components).
> 
> The most related part of the code can be found in:
> 
> [Uni-NaVid/uninavid/model/uninavid_arch.py](https://github.com/jzhzhang/Uni-NaVid/blob/64444eba4d4503d5b08f080673e87229f7fe7f25/uninavid/model/uninavid_arch.py#L502-L519)
> 
> Lines 502 to 519 in [64444eb](/jzhzhang/Uni-NaVid/commit/64444eba4d4503d5b08f080673e87229f7fe7f25)
> 
>  assert video_or_not[cur_image_idx] is True   
>  assert token_idx == 0   
>  assert nav_or_not[cur_image_idx][token_idx].shape[0] == 64   
>  cur_new_input_embeds.append( 
>      self.get_model().embed_tokens(cur_input_ids[:image_token_start])) 
>  seperator_token = self.get_model().embed_tokens(cur_input_ids[image_token_start - 1, None]) 
>  assert not final_token_length_lst[cur_image_idx] is None   
>  video_index = 0 
>  for ii_index, ii in enumerate(final_token_length_lst[cur_image_idx]): 
>      cur_new_input_embeds.append(cur_image_features[video_index:video_index + ii]) 
>      if ii_index == len(final_token_length_lst[cur_image_idx]) - 1: 
>          break 
>      cur_new_input_embeds.append(seperator_token) 
>      video_index += ii 
>  cur_new_input_embeds.append(self.get_model().embed_tokens( 
>      cur_input_ids[image_token_start + 1:image_token_start + 3])) 
>  cur_new_input_embeds.append(nav_or_not[cur_image_idx][token_idx]) 
>   
> You can reorganize the language and visual tokens here, but please note that this requires other related code modifications accordingly.

Thank you so much for the detailed explanation.
I have two follow-up questions regarding the feasibility of using Uni-NaVid for general reasoning tasks **without retraining**:

### 1. Performance vs. Functionality
I understand that changing token patterns typically necessitates retraining for optimal performance. However, I am curious: if I strictly modify the input sequence (e.g., feeding a sequence of images and text for reasoning) *without* retraining, would the model likely fail to generate valid outputs (i.e., crash or output garbage due to pattern mismatch), or is it just a matter of potential performance degradation? 

I am trying to assess if the pre-trained Uni-NaVid has any **zero-shot capability** for general VLM tasks where the strict navigation token structure might not be fully preserved.

### 2. Adapting for Pure Reasoning (VLM Mode)
My current goal is to use Uni-NaVid for a simple reasoning/understanding task (similar to a standard VLM), independent of generating navigation actions.

**Comment by jzhzhang** [comment]
- Yes, just replace the `navigation instruction` tokens with any tokens you prepare, and the model could still output reasonable action outputs. More simply, you can use VLM to caption targets (e.g. an image) into text, and replace the navigation instruction with the text.

- Yes, you can achieve this with Uni-NaVid. You can refer to our data preparation part to see how to prepare ImageQA or VideoQA inputs:
https://github.com/jzhzhang/Uni-NaVid/blob/64444eba4d4503d5b08f080673e87229f7fe7f25/uninavid/train/train.py#L664-L677

**Comment by Yanyx44** [comment]
`    def miletones_summary(self, milestone_images):
            if not milestone_images:
                return ""

            try:
                self.model.get_model().new_frames = len(milestone_images)
                batch_image = np.asarray(milestone_images)
                video_tensor = self.image_processor.preprocess(batch_image, return_tensors='pt')['pixel_values'].half().cuda()
                
                summary_prompt = (
                                "Imagine you are a robot navigator. You have been given a video showing a path you have traveled. "
                                "Your assigned task is: Generate a concise navigation summary for this path."
                            )
                question = summary_prompt.replace(DEFAULT_IMAGE_TOKEN, '').replace('\n', '')
                
                qs = summary_prompt
                if self.model.config.mm_use_im_start_end:
                    qs = DEFAULT_IM_START_TOKEN + DEFAULT_IMAGE_TOKEN + DEFAULT_IM_END_TOKEN + '\n' + qs
                else:
                    qs = DEFAULT_IMAGE_TOKEN + '\n' + qs

                conv = conv_templates[self.conv_mode].copy()
                conv.append_message(conv.roles[0], qs)
                conv.append_message(conv.roles[1], None)
                prompt = conv.get_prompt()

                token_prompt = tokenizer_image_token(prompt, self.tokenizer, IMAGE_TOKEN_INDEX, return_tensors='pt').cuda()
                
                VIDEO_START_SPECIAL_TOKEN = "<video_special>"
                VIDEO_END_SPECIAL_TOKEN = "</video_special>"
                IMAGE_START_TOKEN = "<image_special>"
                IMAGE_END_TOKEN = "</image_special>"
                NAVIGATION_SPECIAL_TOKEN = "[Navigation]"
                IAMGE_SEPARATOR = "<image_sep>"
                image_start_special_token = self.tokenizer(IMAGE_START_TOKEN, return_tensors="pt").input_ids[0][1:].cuda()
                image_end_special_token = self.tokenizer(IMAGE_END_TOKEN, return_tensors="pt").input_ids[0][1:].cuda()
                video_start_special_token = self.tokenizer(VIDEO_START_SPECIAL_TOKEN, return_tensors="pt").input_ids[0][1:].cuda()
                video_end_special_token = self.tokenizer(VIDEO_END_SPECIAL_TOKEN, return_tensors="pt").input_ids[0][1:].cuda()
                navigation_special_token = self.tokenizer(NAVIGATION_SPECIAL_TOKEN, return_tensors="pt").input_ids[0][1:].cuda()
                image_seperator = self.tokenizer(IAMGE_SEPARATOR, return_tensors="pt").input_ids[0][1:].cuda()
                
                indices_to_replace = torch.where(token_prompt == -200)[0]
                
                new_list = []
                while indices_to_replace.numel() > 0:
                    idx = indices_to_replace[0]
                    new_list.append(token_prompt[:idx])
                    new_list.append(video_start_special_token)
                    new_list.append(image_seperator)
                    new_list.append(token_prompt[idx:idx + 1])
                    new_list.append(video_end_special_token)
                    token_prompt = token_prompt[idx + 1:]
                    indices_to_replace = torch.where(token_prompt == -200)[0]
                if token_prompt.numel() > 0:
                    new_list.append(token_prompt)
                input_ids = torch.cat(new_list, dim=0).unsqueeze(0)

                stop_str = conv.sep if conv.sep_style != SeparatorStyle.TWO else conv.sep2
                keywords = [stop_str]
                stopping_criteria = KeywordsStoppingCriteria(keywords, self.tokenizer, input_ids)

                cur_prompt = question
                with torch.inference_mode():
                    self.model.update_prompt([[cur_prompt]])
                    output_ids = self.model.generate(
                        input_ids,
                        images=[video_tensor],
                        do_sample=True,
                        temperature=0.2,
                        max_new_tokens=1024,
                        use_cache=True,
                        stopping_criteria=[stopping_criteria])

                input_token_len = input_ids.shape[1]
                n_diff_input_output = (input_ids != output_ids[:, :input_token_len]).sum().item()
                if n_diff_input_output > 0:
                    print(f'[Warning] {n_diff_input_output} output_ids are not the same as the input_ids')
                outputs = self.tokenizer.batch_decode(output_ids[:, input_token_len:], skip_special_tokens=True)[0]
                outputs = outputs.strip()
                if outputs.endswith(stop_str):
                    outputs = outputs[:-len(stop_str)]
                outputs = outputs.strip()
                
                return outputs`

Thank you for the suggestion, but I followed the video-processing approach and wrote the above code to generate a summary of the trajectory. However, I found that the output still contains only actions.

**Comment by jzhzhang** [comment]
Could you please provide your test cases? I can have a try.

**Comment by Yanyx44** [comment]
> Could you please provide your test cases? I can have a try.

My code might be somewhat complex to test, but my goal is roughly this: in a navigation task—such as a VLN task—at a certain stage, I want to generate a summary of the previous trajectory to monitor the progress.

**Comment by jzhzhang** [comment]
I get your idea. Actually, Uni-NaVid has been trained with this capability (similar to what we did in NaVid: https://pku-epic.github.io/NaVid/).

You can use the prompt:
```You have been given a video of historical observations. Based on this image sequence <video>, please describe the navigation trajectory of the robot.```

However, this is a very challenging task, and the performance is not yet very good.

**Comment by Yanyx44** [comment]
Thank you for your response. I managed to get it working later on as well, though the capability is indeed quite unstable. I don’t have any further questions at the moment. Thanks again for your help.
